File: NodeJS/routes/pades-signature-server-key.js
const express = require('express');
const fs = require('fs');
const path = require('path');
const uuidv4 = require('uuid/v4');
const { PadesSigner, StandardSignaturePolicies } = require('pki-express');

const { Util } = require('../util');
const { PadesVisualElements } = require('../pades-visual-elements');

let router = express.Router();
let appRoot = process.cwd();

router.get('/', (req, res, next) => {

	let outputFile = null;

	// Verify if the provided userfile exists.
	let userfile = path.join(appRoot, 'public', 'app-data', req.query['userfile']);
	if (!fs.existsSync(userfile)) {
		next(new Error('File not found'));
		return;
	}

	// Get an instance of the PadesSigner class, responsible for receiving
	// the signature elements and performing the local signature.
	let signer = new PadesSigner();

	// Set PKI default options (see util.js).
	Util.setPkiDefaults(signer);

	// Set signature policy.
	signer.signaturePolicy = StandardSignaturePolicies.PADES_BASIC_WITH_LTV;

	// Process all independent IO operations in parallel. This should happen
	// before the method "sign" to be called.
	Promise.all([

		// Set PDF to be signed.
		signer.setPdfToSignFromPath(userfile),

		// Set a file reference for the stamp file. Note that file can be
		// referenced later by "fref://{alias}" at the "url" field on the visual
		// representation (see public/vr.json or getVisualRepresentation()
		// method).
		signer.addFileReference('stamp', Util.getPdfStampPath()),

		// Set visual reference. We provide a dictionary that represents the
		// visual representation JSON model.
		signer.setVisualRepresentation(PadesVisualElements.getVisualRepresentation()),

		// The PKCS #12 certificate path.
		signer.setPkcs12FromPath(path.join(appRoot, 'resources', 'Pierre de Fermat.pfx'))

	]).then(() => {

		// Set the certificate's PIN.
		signer.certPassword = '1234';

		// Generate path for output file and add the signature finisher.
		Util.createAppData(); // Make sure the "app-data" folder exists (util.js).
		outputFile = uuidv4() + '.pdf';
		signer.outputFile = path.join(appRoot, 'public', 'app-data', outputFile);

		// Perform the signature.
		return signer.sign();

	}).then(() => {

		// Render the result page.
		res.render('pades-signature-server-key', {
			outputFile: outputFile
		});

	}).catch(err => next(err));

});

module.exports = router;

File: Python/sample/views/pades_signature_server_key.py
"""

This sample performs a local PAdES signature in one step using PKI Express.

"""
import os
import uuid

from flask import render_template
from flask import current_app
from flask import make_response
from flask import Blueprint
from pkiexpress import standard_signature_policies
from pkiexpress import PadesSigner

from sample.utils import set_pki_defaults
from sample.utils import get_pdf_stamp_path
from sample.utils import create_app_data
from sample.utils import get_expired_page_headers
from sample.utils_pades import get_visual_representation

blueprint = Blueprint('pades_signature_server_key', __name__,
                      url_prefix='/pades-signature-server-key')


@blueprint.route('/<userfile>')
def index(userfile):

    try:

        # Verify if the provided userfile exists.
        if not os.path.exists(os.path.join(current_app.config['APPDATA_FOLDER'],
                                           userfile)):
            return render_template('error.html', msg='File not found')

        # Get an instance of the PadesSigner class, responsible for receiving
        # the signature elements and performing the local signature.
        signer = PadesSigner()

        # Set PKI default options (see utils.py).
        set_pki_defaults(signer)

        # Set signature policy.
        signer.signature_policy = \
            standard_signature_policies.PADES_BASIC_WITH_LTV

        # Set PDF to be signed.
        signer.set_pdf_to_sign_from_path(os.path.join(
            current_app.config['APPDATA_FOLDER'], userfile))

        # The PKCS #12 certificate path.
        signer.set_pkcs12_from_path(os.path.join(current_app.static_folder,
                                                 'Pierre de Fermat.pfx'))
        # Set the certificate's PIN.
        signer.cert_password = '1234'

        # Set a file reference for the stamp file. Note that this file can be
        # referenced later by "fref://{alias}" at the "url" field on the visual
        # representation (see content/vr.json or get_visual_representation()
        # method).
        signer.add_file_reference('stamp', get_pdf_stamp_path())

        # Set visual representation. We provide a dictionary that represents the
        # visual representation JSON model.
        signer.set_visual_representation(get_visual_representation())

        # Generate path for output file and add to signer object.
        create_app_data()  # Guarantees that "app data" folder exists.
        output_file = '%s.pdf' % (str(uuid.uuid4()))
        signer.output_file = os.path.join(current_app.config['APPDATA_FOLDER'],
                                          output_file)

        # Perform the signature.
        signer.sign()

        response = make_response(render_template(
            'pades_signature_server_key/index.html', filename=output_file))
        response.headers = get_expired_page_headers()
        return response

    except Exception as e:
        return render_template('error.html', msg=e)
